doc/syslog/python-logger-enhancement.md

```diff
@@ -0,0 +1,152 @@
+# SONiC Python logger enhancement #
+
+## Table of Content
+
+### Revision
+
+### Scope
+
+This document describes an enhancement to SONiC Python logger.
+
+### Definitions/Abbreviations
+
+Log identifier: The identifier of a logger instance. Log message printed by the logger instance usually contains the identifier.
+
+### Overview
+
+SONiC provides two Python logger implementations: `sonic_py_common.logger.Logger` and `sonic_py_common.syslogger.SysLogger`. Both of them do not provide the ability to change log level at real time. Sometimes, in order to get more debug information, developer has to manually change the log level in code on a running switch and restart the Python daemon. This is not convenient.
+
+SONiC also provides a C/C++ logger implementation in `sonic-platform-common.common.logger.cpp`. This C/C++ logger implementation is also a wrapper of Linux standard `syslog` which is widely used by swss/syncd. It provides the ability to set log level on fly by starting a thread to listen to CONFIG DB LOGGER table change. SONiC infrastructure also provides the Python wrapper for `sonic-platform-common.common.logger.cpp` which is `swsscommon.Logger`. However, this logger implementation also has some drawbacks:
+
+
+1. `swsscommon.Logger` assumes redis DB is ready to connect. This is a valid assumption for swss/syncd. But it is not good for a Python logger implementation because some Python script may be called before redis server starting.
```

Comment:
@Junchao-Mellanox How are you overcoming this limitation? I see that you are still writing to the DB which means redis server needs to be running

Reply:
To overcome it, the implementation will:
1. At the beginning of the thread function, it will wait until redis is up. If the connect to redis fails, it will silently ignore the connection error and retry in 5 seconds. 
2. In the while loop of thread function, catch connection error while reading from redis